### PR TITLE
fix(zerocode): use daemon final text when no streaming text was accumulated

### LINKED ISSUE #8644: [Bug]: ZeroCode can complete a Code turn with no visible assistant output
# [Bug]: ZeroCode can complete a Code turn with no visible assistant output

## Affected component

zerocode/tui

## Severity

S2 - degraded behavior

## Current behavior

In a ZeroCode Code session, a submitted message can show the normal working state, complete successfully according to runtime logs, and then return the input bar to idle without rendering any assistant response or diagnostic row.

In the observed case, the Logs pane showed the turn reaching `turn_final_response` after provider response handling completed. The Code transcript still showed the user message, but no new assistant message appeared.

Current source also appears to allow this silent path. At `60b67e094c899b4bb090f2cb888eacc8637b00f6`, `SessionUpdate::TurnComplete { outcome: Completed, content }` passes `content` into `ChatState::commit_turn(content, true)`, but `commit_turn` flushes only accumulated streaming text/thought and then discards the `full_text` argument:

```rust
pub fn commit_turn(&mut self, full_text: String, clean: bool) {
    self.flush_streaming_text();
    self.flush_streaming_thought();
    let _ = full_text;
    ...
}
```

That means `TurnComplete.content` is not a display fallback. If no `AgentMessageChunk` text was accumulated, or if the provider returned a successful empty completion, the completed turn produces no visible transcript entry.

## Expected behavior

ZeroCode should never make a successful turn look like nothing happened.

At minimum:

- If `TurnComplete.content` is non-empty and no streamed text was committed, render that content as the assistant response.
- If the completed turn has no assistant text and no tool output, render an explicit diagnostic such as “Agent returned an empty response.”
- Preserve the existing streaming path without duplicating already-rendered streamed text.

## Steps to reproduce

```bash
1. Start a ZeroCode Code session with a working model provider.
2. Send a follow-up prompt.
3. Wait for the input bar/status to leave the working state.
4. Observe that the transcript can return to idle with no new assistant row.
5. Check Logs and observe the corresponding turn reaches `turn_final_response`.
```

## Impact

Affected users: ZeroCode Code users.

Frequency: intermittent so far; observed during live dogfood use.

Consequence: the user cannot tell whether the model returned an empty response, the TUI dropped the final response, or the turn silently failed. The UI presents a completed turn as a no-op.

## Logs / stack traces

Sanitized live log sequence from the observed turn:

```text
provider/send llm_request
provider/note stream: message_start
provider/note stream: message_delta
provider/note stream: message_stop
provider/receive llm_response
agent/complete turn_final_response
```

The full `turn_final_response` payload was not available through the Logs pane in the observed session, so this issue intentionally focuses on the TUI/runtime behavior that can make an empty or missed final response invisible.

## ZeroClaw version

Observed during live ZeroCode dogfood use on July 3, 2026. Source inspection was against `60b67e094c899b4bb090f2cb888eacc8637b00f6`.

## Rust version

Not relevant to the observed runtime behavior.

## Operating system

macOS.

## Regression?

Unknown.

## Pre-flight checks

- [x] I reproduced this on the latest master branch or latest release.
- [x] I redacted secrets, tokens, and personal data from all submitted content.

## Related follow-ups

This issue is intentionally separate from two related observability/UX gaps discovered while diagnosing it:

- Logs detail can show `Full payload unavailable`, which prevents inspecting `attributes.text` for the turn.
- Doctor can time out with `RPC doctor/run: timed out after 5s` without showing which diagnostic check is blocking.

### PR DESCRIPTION
**Pull request by Project516**
## Summary

- **Base branch:** `master` (all contributions)
- **What changed and why:** `commit_turn` now uses the daemon-provided `TurnComplete.content` as a fallback when no `AgentMessageChunk` streaming text was accumulated during the turn, and renders an explicit diagnostic system message when a completed turn has no assistant text, no tool calls, and no final content. Previously, empty completions were silently discarded, leaving the transcript blank after a successful turn.
- **Scope boundary:** Does not change streaming behavior, tool-call handling, or cancel/failure paths. Only affects the case where a turn completes with no streamed text.
- **Blast radius:** ZeroCode TUI transcript rendering on turn completion. No runtime, provider, gateway, or channel impact.
- **Linked issue(s):** Fixes #8644
- **Labels:** `bug`, `zerocode`, `risk:medium`, `size:S`

### How you can test (when useful)

- **Reviewer testing requested?** N/A — behavior is covered by unit tests; no manual setup needed.
- **Interface(s) exercised:** `tui` — ZeroCode transcript rendering on turn completion
- **Setup / preconditions:** None
- **Steps to run:** N/A
- **Expected on this branch (after):** Empty completions render a diagnostic; non-empty final text renders as fallback; cancelled turns do not show completion diagnostic.
- **Prior behavior on `master` (before):** Empty completions silently vanish from transcript.

### How I tested

```bash
cargo fmt --all -- --check   # clean
cargo clippy -p zerocode --all-targets -- -D warnings   # clean
cargo test -p zerocode   # 443 passed, 0 failed
```

- **CI checks relied on and why they cover this change:** `cargo fmt`, `cargo clippy`, `cargo test -p zerocode` cover the changed chat.rs and locale files
- **Known CI coverage gap, if any:** None
- **Commands run and tail output:** `test result: ok. 443 passed; 0 failed`
- **Beyond CI — what did you manually verified?**
  - `commit_turn_renders_nonempty_fallback_when_no_streaming` — non-empty final text renders once when no streaming text was accumulated
  - `commit_turn_does_not_duplicate_already_flushed_text` — existing test still passes (no duplicate when streaming text was already flushed)
  - `commit_turn_shows_diagnostic_when_no_output_at_all` — empty completion with no streaming, no tools, and no final text renders the diagnostic system message
  - `commit_turn_no_diagnostic_when_not_clean` — cancelled/failed turn (clean=false) with no output does NOT append the completion diagnostic
  - `commit_turn_no_diagnostic_when_tool_calls_present` — empty completion with tool calls present does NOT add diagnostic (tool entries are the visible record)
- **If any command was intentionally skipped, why:** None

## Security & Privacy Impact (required)

- New permissions, capabilities, or file system access scope? No
- New external network calls? No
- Secrets / tokens / credentials handling changed? No
- PII, real identities, or personal data in diff, tests, fixtures, or docs? No
- Prompt injection or untrusted model-visible text introduced/changed? No

## Compatibility (required)

- Backward compatible? Yes
- Config / env / CLI surface changed? No
- Rust/MSRV/toolchain floor changed? No

## Rollback (required for medium/high-risk PRs)

- **Fast rollback:** `git revert <sha>` on the merge commit reverts the chat.rs and locale changes; no data migration needed.
- **Feature flags / toggles:** None — behavior is always-on.
- **Observable failure symptoms if rollback needed:** Users see blank transcripts after turns that produce no streaming text (the pre-existing behavior this PR fixes). Rollback restores that behavior without side effects.

### CODE REVIEW

**Review by Audacity88 — CHANGES_REQUESTED** [review]
@Project516 I reviewed current head `79f434012c9ef7309fa3c2fb4311a928b2b9b87e`, the linked issue #8644, the related already-merged provider-side fix #8663, the current checks, and the exact `apps/zerocode/src/chat.rs` diff. I did not run local Cargo validation or a live ZeroCode smoke in this pass; I treated the posted `cargo fmt`, `cargo clippy -p zerocode`, `cargo test -p zerocode`, and green GitHub CI as the validation record.

I am requesting changes because this fixes the non-empty `TurnComplete.content` fallback, but it still leaves a clean successful turn able to disappear when both streamed text and final content are empty.

### 🔴 Blocking — clean empty completions can still leave no transcript row

#8644 asks ZeroCode to never make a successful turn look like nothing happened. The issue's minimum expected behavior has two parts: use non-empty `TurnComplete.content` when no streamed text was committed, and render an explicit diagnostic when a completed turn has no assistant text and no tool output.

This PR implements the first half, but the second half is still missing. In the new `commit_turn` branch, the fallback only appends an assistant entry when `!self.turn_had_streaming_text && !full_text.is_empty()`. If the daemon sends `TurnComplete { outcome: Completed, content: "" }` after no `AgentMessageChunk` text and no visible tool output, `commit_turn` flushes empty buffers, skips that branch, resets the turn to idle, and leaves the transcript unchanged. That is still the user-visible silent no-op that #8644 was filed to prevent.

Please make that clean no-output completion visible too, for example with a short system/diagnostic row, and add regressions for the three important shapes: no streamed text plus non-empty final text renders once; streamed or pre-tool-flushed text does not duplicate from `full_text`; and clean completion with no streamed text, no final text, and no visible tool output renders the explicit diagnostic. If you want this PR to stay only as the non-empty fallback slice, then the public closing semantics need to change away from `Fixes #8644` so the remaining diagnostic path stays tracked.

### 🟢 What looks good — the duplicate guard is the right shape

The `turn_had_streaming_text` flag is a reasonable per-turn guard for the pre-tool flush case. Setting it when `ToolCall` flushes streamed text, then clearing it at turn start/end, preserves the existing `commit_turn_does_not_duplicate_already_flushed_text` behavior while allowing non-streamed final text to be rendered once.

### 🟡 Warning — fill the medium-risk rollback fields while revising

The body currently labels the PR `risk:medium`, but the rollback section still uses the low-risk template sentence. While updating the PR, please fill the medium-risk fields from the template: fast rollback command/path, feature flags or toggles, and observable failure symptoms. That is public-record cleanup, not the reason for this requested-changes verdict.

**Comment by Project516** [comment]
@Audacity88 Addressed the review feedback:

**Changes pushed:**
- Added `turn_had_tool_calls` flag to distinguish "empty completion with tool calls" from "empty completion with nothing at all"
- Added diagnostic `ChatEntry::SystemMessage("Turn completed with no output.")` when a turn completes with no streamed text, no tool calls, and no final content — prevents the silent no-op you flagged
- Added `zc-turn-no-output` translation key to all 5 locales (en, zh-CN, es, ja, fr)
- Added 3 regression tests covering all three shapes: non-empty fallback renders once, diagnostic renders for empty completion, no diagnostic when tools present
- Filled in medium-risk rollback section (fast rollback, feature flags, observable symptoms)

**Validation:** `cargo fmt` clean, `cargo clippy -p zerocode --all-targets -- -D warnings` clean, `cargo test -p zerocode` 439 passed, 0 failed.

Please re-review when ready.

**Review by Audacity88 — APPROVED** [review]
@Project516

Reviewed current head `fbeaed7`. I checked the live PR state, the updated diff, issue #8644, the merged provider-side fix #8663, my prior review, the author update comment, current CI, and the new ZeroCode regressions. I did not run local Cargo validation or a live ZeroCode smoke in this review-only pass; I am treating the author-reported focused ZeroCode commands, the targeted state-machine tests, and the now-passing GitHub checks as the code-review evidence for clearing my prior block.

### ✅ Resolved — Empty successful turns now get a visible transcript row

This update covers the missing half from my prior review. `commit_turn` now uses the daemon-provided final text when no streaming text was committed, and it also renders a localized system diagnostic when a clean completed turn has no streamed text, no final text, and no tool call record. That matches #8644's minimum behavior: a successful turn should not return the input bar to idle with no visible transcript change.

The regression coverage now covers the shapes I asked for: the new tests cover non-streamed final text, empty/no-tool diagnostics, and tool-call-only quiet behavior, while the existing duplicate-guard test still covers already-flushed streamed text.

### ✅ Resolved — The rollback and issue-closing story now match the scope

The PR body now describes both the non-empty final-content fallback and the explicit diagnostic path, so `Fixes #8644` is no longer overclaiming the issue. The rollback section is also filled in with the command/path, toggle state, and observable failure symptom, which addresses my earlier public-record warning.

### 🟢 What looks good — The per-turn state stays narrow

The added flags are scoped to the active turn and reset at turn start/end, which keeps the duplicate guard local to the transcript commit path. That preserves the existing pre-tool text behavior while making the no-output completion visible to the user.

This clears my prior request-changes review on the code. It is not a live-smoke claim; I did not manually exercise the TUI path in this pass.

File: apps/zerocode/src/chat.rs [hunk 1]
```diff
@@ -5802,9 +5820,30 @@ impl ChatState {
...
         self.flush_streaming_thought();
-        let _ = full_text;
+        // If no streaming text was accumulated during this turn, use the
+        // daemon-provided final text as a fallback so the turn is never
+        // invisible to the user.
+        if !self.turn_had_streaming_text && !full_text.is_empty() {
+            self.entries
+                .push(ChatEntry::AgentMessage(Arc::<str>::from(full_text)));
+            self.mark_dirty_append();
+        } else if !self.turn_had_streaming_text && !self.turn_had_tool_calls && full_text.is_empty()
```
Comment (IftekharUddin) [inline apps/zerocode/src/chat.rs]: ### 🔴 Blocking — gate this diagnostic on clean completion

This branch also runs when `clean == false`. Cancelled/failed daemon outcomes first append their attributed reason and then call `commit_turn(String::new(), false)`; cancel-watchdog and local cancel-error paths do the same. With no streamed text or tool call, they now receive this success-only “Turn completed with no output” row as well. Please require `clean` here (or separate the settlement paths) and add a non-clean regression so cancelled/failed turns retain only their appropriate diagnostic.

**Review by IftekharUddin — CHANGES_REQUESTED** [review]
@Project516 I reviewed current head 02cc816b72a43f901dada5809c46856613ba53b8, the linked issue, complete diff, turn-end call sites, prior reviews, and green required checks. I found one cancellation/failure regression that is not covered by the existing approval.

### 🔴 Blocking — the success diagnostic also fires for cancelled and failed turns

`commit_turn` is shared by both clean completion and non-clean settlement. Every cancellation/failure path calls `commit_turn(String::new(), false)` after recording its specific reason (or after a local cancel error/watchdog path). The new diagnostic branch ignores `clean`, so an empty non-clean turn with no tool call appends “Turn completed with no output.” as though it completed successfully.

```mermaid
flowchart LR
    C["Completed"] -->|"clean = true"| D{"no text or tools?"}
    X["Cancelled / Failed"] -->|"clean = false"| R["render attributed reason"]
    R --> D
    D -->|"current code: yes"| M["Turn completed with no output"]
```

The diagnostic must be limited to the clean-completion branch. Please gate it on `clean` (or split the clean and non-clean settlement paths) and add a regression proving `commit_turn(String::new(), false)` does not append the success diagnostic. A `TurnEndOutcome::Cancelled`/`Failed` test that preserves only the attributed system reason would pin the user-visible behavior most directly.

### 🟢 What looks good — the clean fallback cases are otherwise covered

For a genuinely completed turn, the final-text fallback and empty-output diagnostic match #8644. The duplicate guard preserves pre-tool streamed text, the tool-only case remains visible without a redundant diagnostic, all five locale keys are present, and the current required suite is green.

Requesting changes for the non-clean-path regression above.

**Comment by Project516** [comment]
Thanks for the detailed review. Branch rebased onto current `master`.

**Actioning your feedback:**

1. **Diagnostic fires for cancelled/failed turns** — You're right. `commit_turn(String::new(), false)` is called on cancellation/failure paths, and the new diagnostic branch ignores the `clean` flag. I'll gate the "Turn completed with no output" diagnostic on `clean == true` so it only fires for genuinely completed turns. Cancelled/failed turns will continue to render their attributed reason only.

2. **Clean empty completions still leaving no transcript row** — I'll add a fallback that renders a short diagnostic entry when a clean turn has no streamed text, no final text, and no visible tool output.

3. **Rollback fields** — Will fill the medium-risk rollback section (fast rollback: `git revert <sha>`, feature toggle: N/A, observable symptoms: empty transcript entries on turn completion).

Will update the PR with the fixes.

**Comment by Project516** [comment]
Thanks for the review, @IftekharUddin. Your blocking finding is already addressed in the latest push:

**Gated on `clean`** — The diagnostic branch is `else if clean && !self.turn_had_streaming_text && !self.turn_had_tool_calls && full_text.is_empty()`. Cancelled/failed turns (`clean == false`) will never emit the "Turn completed with no output" row. They continue to render only their attributed reason (cancel/timeout/failure).

**Regression test** — `commit_turn_no_diagnostic_when_not_clean` creates a state with `turn_in_flight = true`, calls `commit_turn(String::new(), false)`, and asserts `entries.is_empty()` — proving cancelled turns produce no diagnostic.

All 4 regression tests pass:
- `commit_turn_renders_nonempty_fallback_when_no_streaming` — fallback renders once
- `commit_turn_shows_diagnostic_when_no_output_at_all` — diagnostic fires for empty clean completion
- `commit_turn_no_diagnostic_when_not_clean` — no diagnostic for cancelled/failed turns
- `commit_turn_no_diagnostic_when_tool_calls_present` — no diagnostic when tools were called

PTAL when convenient.

**Review by Audacity88 — APPROVED** [review]
@Project516, I re-reviewed current head `c83dda201f9719f3271fa8818593ccbe2b47ce23` against the full six-file diff, #8644, my prior reviews, Iftekhar's active blocker, your latest responses, every `commit_turn` call site, the current required checks, and a clean synthetic merge with current `master`. I did not rerun local Cargo or a live ZeroCode session.

### ✅ Resolved — non-clean turns no longer receive the success diagnostic

The no-output diagnostic is now gated on `clean`. The focused regression directly proves that an empty `commit_turn(..., false)` settlement appends no “Turn completed with no output” row, matching the non-clean contract used by the current callers. That resolves the remaining helper-level production concern raised in Iftekhar's review; their active `CHANGES_REQUESTED` remains theirs to clear.

### 🟢 What looks good — the turn-output matrix is now coherent

The focused coverage now distinguishes five important shapes: non-streamed final text renders once, already-flushed streamed text is not duplicated, a clean textless/tool-less completion gets a diagnostic, a tool-only turn does not get that diagnostic, and a non-clean settlement does not get it either. The two per-turn flags are reset at both turn start and settlement, keeping the fallback state local to one turn.

### 🟡 Warning — refresh and remove source-history references before merge

The branch merges structurally with current `master`, but it is behind changes that added the source-comment hygiene gate and touched `chat.rs` plus all five locale files. Two added Rust comments still mention `#8644`; current policy rejects issue or PR references in source comments. Please refresh the branch, rewrite those comments around the durable invariant, and let required CI rerun before merge. This approval is for the reviewed code on `c83dda20`, not a merge-readiness claim.

**Comment by Project516** [comment]
Rebased onto current \`master\` (no conflicts).

@IftekharUddin the cancellation/failure-path finding is already fixed in the head you reviewed on 07-14 and confirmed by my 07-15 reply: the diagnostic branch in \`commit_turn\` is gated on \`clean\`, so \`commit_turn(String::new(), false)\` on cancel/failure paths never emits it. Re-ran the full \`commit_turn\` test module locally (5/5 pass, including \`commit_turn_no_diagnostic_when_not_clean\`) to confirm nothing regressed in the rebase. Flagging in case this just needs a re-review pass to clear.

**Review by IftekharUddin — APPROVED** [review]
@Project516 I re-reviewed current head 43d344ae85cca8a2d99168546a567cd90ba634f7 against #8644, merged provider-side hardening #8663, the complete review history, every current `commit_turn` call shape, the full ZeroCode/localization diff, and the current exact-head checks. I did not rerun Cargo or perform a live TUI smoke in this pass.

### ✅ Resolved — non-clean settlements no longer receive the success diagnostic

My prior blocker is addressed. The no-output diagnostic now requires `clean`, and `commit_turn_no_diagnostic_when_not_clean` pins the helper contract used by cancellation, failure, local cancel-error, and watchdog settlement paths. Those paths continue to render their attributed reason without adding “Turn completed with no output.”

### ✅ Resolved — source-comment hygiene is clean after the refresh

The branch refresh removed the issue-number references from the added Rust comments. The comments now describe the durable streaming/tool-call invariants, and the Repository Structure and Docs Style checks have passed on the current head.

### 🟢 What looks good — the visible-output matrix matches the bug contract

The current code distinguishes the five relevant outcomes: non-streamed final text renders once; streamed or pre-tool-flushed text is not duplicated; a clean textless/tool-less completion gets a localized diagnostic; a tool-only completion remains visible without a redundant diagnostic; and a non-clean settlement does not claim success. The two added booleans are authoritative ephemeral facts for the active turn and reset at both admission and settlement, so they do not duplicate daemon or persisted state. All locales registered in `locales.toml` carry the diagnostic key.

### 🟢 What looks good — exact-head required CI is green

The complete required matrix, including Test, Lint, the platform and feature checks, benchmarks compile, Installer Drift, and CI Required Gate, has now finished successfully on unchanged head 43d344ae85cca8a2d99168546a567cd90ba634f7. I found no remaining code, review-history, or public-record blocker, so this is ready for approval.
